benchexec/tools/ultimatetestgenerator.py

```diff
@@ -0,0 +1,155 @@
+                return "ERROR: EXCEPTION"
+            if self._contains_overapproximation_result(line):
+                return "UNKNOWN: OverapproxCex"
+            if termination_false_string in line:
+                return result.RESULT_FALSE_TERMINATION
+            if termination_true_string in line:
+                return result.RESULT_TRUE_PROP
+            if ltl_false_string in line:
+                return "FALSE(valid-ltl)"
+            if ltl_true_string in line:
+                return result.RESULT_TRUE_PROP
+            if unsafety_string in line:
+                return result.RESULT_FALSE_REACH
+            if mem_deref_false_string in line:
+                return result.RESULT_FALSE_DEREF
+            if mem_deref_false_string_2 in line:
+                return result.RESULT_FALSE_DEREF
+            if mem_free_false_string in line:
+                return result.RESULT_FALSE_FREE
+            if mem_memtrack_false_string in line:
+                return result.RESULT_FALSE_MEMTRACK
+            if overflow_false_string in line:
+                return result.RESULT_FALSE_OVERFLOW
+            if safety_string in line or all_spec_string in line:
+                return result.RESULT_TRUE_PROP
+            if treeautomizer_unsat in line:
+                return "unsat"
+            if treeautomizer_sat in line or all_spec_string in line:
+                return "sat"
+
+        return result.RESULT_UNKNOWN
+
+    @staticmethod
+    def _determine_result_with_property_file(run):
+        for line in run.output:
+            if line.startswith("DONE"):
+                return result.RESULT_DONE
+            elif line.startswith("FALSE(valid-free)"):
+                return result.RESULT_FALSE_FREE
+            elif line.startswith("FALSE(valid-deref)"):
+                return result.RESULT_FALSE_DEREF
+            elif line.startswith("FALSE(valid-memtrack)"):
+                return result.RESULT_FALSE_MEMTRACK
+            elif line.startswith("FALSE(valid-memcleanup)"):
+                return result.RESULT_FALSE_MEMCLEANUP
+            elif line.startswith("FALSE(TERM)"):
+                return result.RESULT_FALSE_TERMINATION
+            elif line.startswith("FALSE(OVERFLOW)"):
+                return result.RESULT_FALSE_OVERFLOW
+            elif line.startswith("FALSE"):
+                return result.RESULT_FALSE_REACH
+            elif line.startswith("TRUE"):
+                return result.RESULT_TRUE_PROP
+            elif line.startswith("UNKNOWN"):
+                return result.RESULT_UNKNOWN
+            elif line.startswith("ERROR"):
+            	if not line.startswith("ERROR: Caught known exception: Unsupported non-linear arithmetic"):
+                    status = result.RESULT_ERROR
+                    if line.startswith("ERROR: INVALID WITNESS FILE"):
+                        status += " (invalid witness file)"
+                    return status
+        return result.RESULT_UNKNOWN
```

Comment:
It seems that this is a lot of code copied from `ultimate.py`, with some small changes.

Do (some of) those change make sense to have in `ultimate.py` in general? I guess for example that this is the case for the check for non-linear arithmetic, or at least it wouldn't hurt to have in `ultimate.py`?

Where changes cannot be upstreamed: Can those methods be overwritten in a way that most of the code does not need to be copied, but is called with super?

Reply:
I dont want to iterate twice over the logfile and i think it is better if I return the non-linear arithmetic error anyway. I removed return values that are not used in test-case generation.

Reply:
But this still uses duplicate code, and if Ultimate changes its output, then both `ultimate.py` and the new code here needs to be updated, which is easily forgotten. Can't there be one method that supports parsing the output for all the use cases?

Reply:
I changed `ultimate.py`, Daniel gave his ok with this change. Any further suggestions?

Reply:
Thanks, this is already much nicer, partially because now we can see what the actual difference is.

And it seems that the only difference in output parsing is the additional `line.startswidth("DONE")`. So I am wondering, would it really hurt to move that condition up into `ultimate.py`? It doesn't look at if this would break parsing other Ultimate variant outputs, and it would simplify the whole code even more. You could also choose to make the check stricter than `startswith("DONE")` if you fear it might match something else accidentally.

Reply:
You are right, @danieldietsch and I decided to change `ultiamte.py` back and added the condition there.

Reply:
Nice! Thanks being willing to iterate on this with me. I am confident that the current solution will create less maintenance effort in the future.